### ISSUE TITLE
Feature Request - Ability to bind Workspaces

### ISSUE THREAD

**Issue by davissp14** [tektoncd/dashboard#1283]
It would be really nice to be able to configure a workspace prior to creating a PipelineRun within the dashboard.

I understand that triggers can be used to do this automatically through events, but for folks who want to manually kick something off through the dashboard, this would be pretty useful.

**Comment by AlanGreene** [comment]
I agree it would be nice to be able to manually create runs using workspaces.

We're currently working on introducing the ability to manually create a TaskRun, similar to the existing functionality for PipelineRuns. As part of this we're likely to revisit some of the PipelineRun pieces.

Some initial thoughts for anyone interested in picking this up...

Adding support for workspaces would require:
- knowledge of available PVCs, ConfigMaps, changes to how we handle Secrets
  - would it be sufficient to list existing resources of these types without providing the ability to create / edit them?
- similar logic as we have for resources / params to detect the required resources and present them in the create UI
  - given the potential for many more items in this UI we really should look at moving to the full page experience instead of the current modal dialog, see item 5 of https://github.com/tektoncd/dashboard/issues/966
- possibly more...

Reference:
- [Workspaces docs](https://github.com/tektoncd/pipeline/blob/master/docs/workspaces.md). More info in TaskRun and PipelineRun docs that may be of use.
- [TaskRun example](https://github.com/tektoncd/pipeline/blob/master/examples/v1beta1/taskruns/workspace.yaml)
- [PipelineRun example](https://github.com/tektoncd/pipeline/blob/master/examples/v1beta1/pipelineruns/workspaces.yaml)

**Comment by davissp14** [comment]
It seems a little silly that workspace definitions happen on the PipelineRun/TaskRun resource given how they are used.  Basically, you'd have to specify one-to-many workspaces, per run, and hope that the end-user is able to map the bindings correctly on each go.  It doesn't look like there' a reliable way to determine the workspace specifications/requirements by reading the pipeline.

Reference: https://github.com/tektoncd/pipeline/blob/master/docs/pipelineruns.md

> Creation of a PipelineRun will trigger the creation of TaskRuns for each Task in your pipeline.

It seems like this kind of configuration should belong on the Pipeline, or there needs to be another level of abstraction "PipelineTemplate" that can hold re-usable configuration data that PipelineRun's can inherit.

**Comment by AlanGreene** [comment]
Depending on your specific use cases, if you're looking for a way to template the creation of runs you may be interested in [Tekton Triggers](https://github.com/tektoncd/triggers/tree/master/docs).

> Triggers enables users to map fields from an event payload into resource
templates. Put another way, this allows events to both model and instantiate
themselves as Kubernetes resources. In the case of `tektoncd/pipeline`, this
makes it easy to encapsulate configuration into `PipelineRun`s and
`PipelineResource`s.

Changes to how Pipelines / Tasks are defined would need to raised against the core [Tekton Pipelines](https://github.com/tektoncd/pipeline) and likely raised for discussion at the weekly working group meeting. Since Pipelines and Tasks are intended to be reusable, configuration such as workspaces, service accounts, resources, etc. are considered `runtime` options. See https://github.com/tektoncd/pipeline/issues/2141 and https://github.com/tektoncd/pipeline/issues/2140#issuecomment-593892029.

**Comment by davissp14** [comment]
I actually opened up a related issue shortly after I opened up this one.  Thank you!

**Comment by davissp14** [comment]
Here's a link to the related issue: https://github.com/tektoncd/pipeline/issues/2372

**Comment by AlanGreene** [comment]
Moving the Create TaskRun UI to a full-page experience as we've already done for Create PipelineResource and Create Secret should be happening very soon: https://github.com/tektoncd/dashboard/issues/966

Once this is done we'll have more space to work with and may make some further layout changes to the existing Create PipelineRun / TaskRun UI. These would allow us to provide a cleaner and more user friendly experience, especially as we start adding more functionality, e.g. https://github.com/tektoncd/dashboard/issues/1722

**Comment by AlanGreene** [comment]
This is still something we want to add, freezing so it doesn't get automatically closed.

/lifecycle frozen

**Comment by morningspace** [comment]
Any update for this issue? Really looking forward to this feature. Without specifying workspace, it looks the manually kicking off PipelineRun or TaskRun feature is not really useful in many cases.

**Comment by jasperjonker** [comment]
I am also curious, would be extremely helpful to trigger pipelines from the dashboard

**Comment by marniks7** [comment]
Hi,
- [List](https://tekton.dev/docs/pipelines/workspaces/#specifying-volumesources-in-workspaces) of sources to support: PVC, ConfigMap, Secrets, `emptyDir`, VolumeClaimTemplate, projected (alpha), csi (alpha).
- Features: [subPath](https://tekton.dev/docs/pipelines/workspaces/#persistentvolumeclaim) for PVC, 
What the use cases for subPath we have? not sure how users should understand what to put there

### Basic Solution
 PVC, ConfigMap, Secrets, `emptyDir` only
- Basic solution is to show dropdown with names of configMaps, Secrets and PVC
![image](https://user-images.githubusercontent.com/9289172/188518856-bc33ecc3-e855-4dd5-be73-8ad2536d2098.png)

- Also, it is possible to add types before names, like `secret: default-token-6q424`, `configmap: kube-root-ca.crt`

- I am surprised to see `router.go` as api proxy for kubernetes API(s) allowing to list everything with related clusterrole access. In order to work with secrets, api should expose only required endpoints without providing access to secrets itself. Unfortunately, there is no RBAC access in k8s to read only secret metadata. (`list` access will allow to read everything). As of now idk how to solve this problem on security level... admins will need to give `list` access for `secrets` to `tekton-dashboard-backend`.

### subPath
- `subPath` (but should be displayed only for PVC) : 
![image](https://user-images.githubusercontent.com/9289172/188521413-278319a0-8a9d-4280-a525-94b2d478d24f.png)

### More complex
VolumeClaimTemplate, projected (alpha), csi (alpha) not analyzed

**Comment by AlanGreene** [comment]
> I am surprised to see router.go as api proxy for kubernetes API(s) allowing to list everything with related clusterrole access. In order to work with secrets, api should expose only required endpoints without providing access to secrets itself.

Yes we would need to either apply a filter to the Secret endpoints to return only the basic metadata, provide a custom endpoint specifically for this purpose (and block the default endpoint), or provide free-text input for the Secret name in the UI instead of providing a dropdown list from which the user would select.

**Comment by gsgs-libin** [comment]
> Hi,
> 
> * [List](https://tekton.dev/docs/pipelines/workspaces/#specifying-volumesources-in-workspaces) of sources to support: PVC, ConfigMap, Secrets, `emptyDir`, VolumeClaimTemplate, projected (alpha), csi (alpha).
> * Features: [subPath](https://tekton.dev/docs/pipelines/workspaces/#persistentvolumeclaim) for PVC,
>   What the use cases for subPath we have? not sure how users should understand what to put there
> 
> ### Basic Solution
> PVC, ConfigMap, Secrets, `emptyDir` only
> 
> * Basic solution is to show dropdown with names of configMaps, Secrets and PVC
>   ![image](https://user-images.githubusercontent.com/9289172/188518856-bc33ecc3-e855-4dd5-be73-8ad2536d2098.png)
> * Also, it is possible to add types before names, like `secret: default-token-6q424`, `configmap: kube-root-ca.crt`
> * I am surprised to see `router.go` as api proxy for kubernetes API(s) allowing to list everything with related clusterrole access. In order to work with secrets, api should expose only required endpoints without providing access to secrets itself. Unfortunately, there is no RBAC access in k8s to read only secret metadata. (`list` access will allow to read everything). As of now idk how to solve this problem on security level... admins will need to give `list` access for `secrets` to `tekton-dashboard-backend`.
> 
> ### subPath
> * `subPath` (but should be displayed only for PVC) :
>   ![image](https://user-images.githubusercontent.com/9289172/188521413-278319a0-8a9d-4280-a525-94b2d478d24f.png)
> 
> ### More complex
> VolumeClaimTemplate, projected (alpha), csi (alpha) not analyzed

In my pipeline, different tasks use different workspaces and serviceaccounts, so hopefully UI will be able to support it as well.
like this:
```shell
      apiVersion: tekton.dev/v1beta1
      kind: PipelineRun
      metadata:
        generateName: my-pipelinerun-
        namespace: tekton-run
      spec:
        taskRunSpecs:
          - pipelineTaskName: fetch-ops-git
            taskServiceAccountName: sa-gitlab-clone
          - pipelineTaskName: fetch-app-git
            taskServiceAccountName: sa-gitlab-clone
          - pipelineTaskName: azure-git-action
            taskServiceAccountName: sa-azure-git-action
        pipelineRef:
          name: php-pipeline-pre
        workspaces:
          - name: git-clone-workspace
            persistentVolumeClaim:
              claimName: pvc-tekton
            subPath: pvc-$(uid)
          - name: image-secret-workspace
            secret:
              secretName: swr-image-secret
          - name: azure-git-action-workspace
            emptyDir: {}
```

**Comment by marniks7** [comment]
With new YAML mode any user-specific cases are supported now, and we may continue discussion.

@AlanGreene, I agree that secrets should not be allowed to read by dashboard UI for those who are concerned with security and we can support both options:
1. Provide ability to specify secret name without dropdown (less user friendly, since user will need to know somehow the name, but better for those concerned about security)
2. In case if access to secrets provided for _specific namespace_ and _specific user_ - show dropdown

**Comment by gsgs-libin** [comment]
Been looking forward to this feature for a long time

**Comment by gsgs-libin** [comment]
> Been looking forward to this feature for a long time

YAML mode, no problem for SRE, but compared to ordinary R&D personnel, the threshold is a bit high, they just want to perform simple operations through the dashboard.

**Comment by ileixe** [comment]
+1 from user perspective. This would be an import issue to use dashboard as frontend workflow while the workspace is the very first primitive for pipeline.

**Comment by fpt-dylan-datnt305** [comment]
+1 here. The issue seems to exists for a long time but no possible solution seems to be addressed. Currently running pipeline from UI always encounter error due to lacking of workspace. YAML mode can be an alternative but not user-friendly and require much overhead

**Comment by jameshwc** [comment]
Could this be prioritize, or could you guide me on which part of codebase should I look at if I want to contribute? This is by far the most painful point we're encountering with dashboard

**Comment by techrc** [comment]
Five years gone，the issue has not been solved yet, tekton maybe dead.

**Comment by AlanGreene** [comment]
@techrc I'm sorry to hear you feel that way, however that's not the case. This feature just isn't a high priority for me compared to other items at this time.

I have had extensive discussions with a number of people who were interested in contributing this, working with them to help define and refine the scope and changes required. These discussions have happened here on GitHub, on Slack, and over Zoom. You can find some of them linked above, including even some PRs with POCs or partial implementations.

Ultimately, none of these people followed through to deliver the functionality, whether due to lack of time, change in priorities, or other reasons. Contributions are still very much welcome though.
